asadiqbal08/WL-272 - Studio Language Selection

asadiqbal08 opened the pull request:
[WL-272](https://openedx.atlassian.net/browse/WL-272)
As a course team user, when I am using Studio I want to be able to change the language that Studio displays so I can interact with the system using a language that I understand.

asadiqbal08 commented:
@mattdrayer @ziafazal FYI.

mattdrayer commented on cms/envs/common.py:
```diff
@@ -190,6 +190,9 @@
 
     'ORGANIZATIONS_APP': False,
+
+    # Show LMS Language selector
```
This is not an LMS feature flag -- also the comment is somewhat redundant -- I would either change it to provide more context or remove it.

mattdrayer commented on cms/envs/common.py:
```diff
@@ -328,6 +331,9 @@
     'django.contrib.messages.middleware.MessageMiddleware',
     'track.middleware.TrackMiddleware',
 
+    # Allows us to set user preferences
```
This comment seems like it does not describe what the middleware is doing

mattdrayer commented on cms/templates/widgets/header.html:
```diff
@@ -189,13 +190,13 @@ <h2 class="info-course">
 
     <div class="wrapper wrapper-r">
       % if user.is_authenticated():
+      <% is_user_authenticated = True %>
```
Maybe I'm missing it but I can't see the usage of this parameter?

mattdrayer commented on lms/templates/navigation.html:
```diff
@@ -180,6 +159,34 @@ <h2 class="course-header"><span class="provider">${course.display_org_with_defau
+              <form action="/i18n/setlang/" method="post" class="settings-language-form" id="language-settings-form">
+                <input type="hidden" id="csrf_token" name="csrfmiddlewaretoken" value="${csrf_token}">
+                % if is_user_authenticated:
+                <input title="preference api" type="hidden" id="preference-api-url" class="url-endpoint" value="${reverse('preferences_api', kwargs={'username': user.username})}" data-user-info="${is_user_authenticated}">
```
It would be better to rename data-user-info to something that better-describes the information it contains -- such as data-user-authenticated, or data-user-auth-status, etc.

mattdrayer commented on lms/templates/navigation.html:
```diff
@@ -180,6 +159,34 @@ <h2 class="course-header"><span class="provider">${course.display_org_with_defau
       </%block>
     </ol>
     % endif
+    % if static.show_language_selector():
```
Is this code block a duplicate of what is present in header.html?  If so is there any way we can combine the two into a reusable include or something?

asadiqbal08 replied:
@mattdrayer header.html and navigation.html are two different files, one is of LMS while other is related to Studio but both are using the common static content file.

mattdrayer commented on cms/templates/widgets/header.html:
```diff
@@ -231,6 +232,33 @@ <h2 class="sr">${_("Account Navigation")}</h2>
       </nav>
       % endif
+        % if static.show_language_selector():
+        <% languages = released_languages() %>
```
Can we move the call to released_languages into the static library?

mattdrayer commented on cms/templates/widgets/header.html:
```diff
@@ -189,13 +190,13 @@ <h2 class="info-course">
 
     <div class="wrapper wrapper-r">
       % if user.is_authenticated():
```
Can we move the user authentication check into the static library?  (assuming it makes sense)

mattdrayer commented on common/djangoapps/lang_pref/views.py:
```diff
@@ -0,0 +1,23 @@
+
+
+@csrf_exempt
+def update_language_session(request):
```
Let's rename this to "update_session_language"

mattdrayer commented on common/djangoapps/lang_pref/views.py:
```diff
@@ -0,0 +1,23 @@
+    Update the language session key.
+    """
+    if request.method == 'PATCH':
+        # Setting the session language to the browser language, if it is supported.
```
We don't really know where the language value comes from, so we should not assume it is coming from the browser.

mattdrayer commented on common/static/js/fixtures/lang-edx-fixture.html:
```diff
@@ -1,5 +1,5 @@
 <form action="/i18n/setlang/" method="post" class="settings-language-form" id="language-settings-form">
-    <input title="preference api" type="hidden" id="preference-api-url" value="/api/user/v1/preferences/test1/">
+    <input title="preference api" class="url-endpoint" type="hidden" id="preference-api-url" value="/api/user/v1/preferences/test1/" data-user-info="True">
```
See below comment about renaming this attribute

mattdrayer commented on common/static/js/src/lang_edx.js:
```diff
@@ -18,32 +16,49 @@ var edx = edx || {},
-                    var language = this.value;
+                    var language = this.value,
+                        url = $('.url-endpoint').val(),
+                        is_user_authenticated =  JSON.parse($('.url-endpoint').data('user-info').toLowerCase());
```
See comment about renaming this attribute

mattdrayer commented on lms/urls.py:
```diff
@@ -103,6 +103,9 @@
     url(r'^api/organizations/', include('organizations.urls', namespace='organizations')),
+
+    # Update session view
+    url(r'^update_language_session$', 'lang_pref.views.update_language_session'),
```
Is there a prefix we should consider for this endpoint?  "api/", "lang-pref/", etc.?  It seems odd for this to be a top level resource.   Also it would be nicer if we could model this as a resource-oriented endpoint, such as "session_language".

asadiqbal08 commented:
@mattdrayer I have incorporated your suggestion and updated the PR. I have made some changes in .scss files BUT It is good if any CSS expert can review the style changes in this PR.
@ziafazal please review the PR. 

Thanks

asadiqbal08 commented:
Jenkins run bokchoy

asadiqbal08 commented:
jenkins run bokchoy

asadiqbal08 commented:
jenkins run bokchoy

mattdrayer commented on common/djangoapps/lang_pref/views.py:
```diff
@@ -0,0 +1,23 @@
+    Update the language session key.
+    """
+    if request.method == 'PATCH':
+        # User select language from the language selector and setting in the language session.
```
I think we can remove this comment

ziafazal commented on common/djangoapps/lang_pref/middleware.py:
```diff
@@ -21,20 +22,33 @@ def process_request(self, request):
+                        if browser_lang in system_released_languages:
+                            language = unicode(browser_lang)
+                            break
+                    set_user_preference(request.user, LANGUAGE_KEY, language)
```
@asadiqbal08 is decided now that we are going to sett user's preferred language automatically and it should not be explicitly set by user?

asadiqbal08 replied:
I thing yes, I had a conversation with @mattdrayer 

I was thinking it could be possible that a language might be removed from the list of "supported languages" for some reason -- perhaps the translations become corrupted, or another operator chooses to remove a language for some other valid reason.  In this case the user's preference would no longer be valid.
So we could then consider the browser language setting, if it is supported by the system.
And if that is also unsupported, we could fall back onto the system default language

andy-armstrong replied:
I disagree with doing this, because it isn't the user's preference. If they log in using a browser on a different machine, it shouldn't set their preference going forward. The browser's language is just a useful default if we are unable to respect the user's preference (or if they haven't set one). I also don't see the benefit because the language will already be correct within the request's session.

asadiqbal08 replied:
@andy-armstrong here we are setting the language IF the user preferred language has been removed from system "supported languages". @mattdrayer let me know is it required or not ?

mattdrayer replied:
This is not a required piece of logic, but as @asadiqbal08 indicated we were wondering what to do in the case of a user preference setting for a language which was no longer supported by the system -- in this case we're trying to degrade gracefully by attempting to apply the browser language.

andy-armstrong replied:
I'd rather clear their preference than set it to something that might not be their preference.

mattdrayer replied:
I'm okay with that -- @asadiqbal08 can you update the code so that it removes an invalid user language preference instead of setting it to the browser language?

ziafazal commented on common/djangoapps/pipeline_mako/templates/static_content.html:
```diff
@@ -132,3 +133,11 @@
     return settings.FEATURES.get('SHOW_LANGUAGE_SELECTOR', False)
 %></%def>
+
+<%def name="is_user_authenticated()"><%
```
I don't think we should write a separate method here when we can do it like this in templates `user.is_authenticated()`

mattdrayer replied:
Ah, yeah, that was something I suggested -- I think I maybe have gotten overzealous with our use of static -- it seems generally-acceptable to reference the user object directly in templates, so I am okay reverting back to using it.

asadiqbal08 commented:
jenkins run bokchoy

ziafazal commented on common/static/js/src/lang_edx.js:
```diff
@@ -1,12 +1,10 @@
 var edx = edx || {},
```
What is thre reason for defining `edx` here?

ziafazal commented on common/static/js/fixtures/lang-edx-fixture.html:
```diff
@@ -1,5 +1,5 @@
 <form action="/i18n/setlang/" method="post" class="settings-language-form" id="language-settings-form">
-    <input title="preference api" type="hidden" id="preference-api-url" value="/api/user/v1/preferences/test1/">
+    <input title="preference api" class="url-endpoint" type="hidden" id="preference-api-url" value="/api/user/v1/preferences/test1/" data-user-is-authenticated="true">
```
Where we are using this fixture or what is the reason for changes in this file?

asadiqbal08 replied:
We are using this in `common/static/js/spec/lang_edx_spec.js` , These tests were written to support WL-243 (Language selector in LMS). and  `common/static/js/src/lang_edx.js` is using the `url-endpoint` class

ziafazal commented on cms/templates/widgets/header.html:
```diff
@@ -188,14 +187,39 @@ <h2 class="info-course">
+          <form action="/i18n/setlang/" method="post" class="settings-language-form" id="language-settings-form">
+              <input type="hidden" id="csrf_token" name="csrfmiddlewaretoken" value="${csrf_token}">
+              % if static.is_user_authenticated():
+              <input title="preference api" type="hidden" id="preference-api-url" class="url-endpoint" value="${reverse('preferences_api', kwargs={'username': user.username})}" data-user-is-authenticated="true">
```
Let's remove the `id` attributes of these hidden inputs if they are no longer used.

ziafazal commented on common/djangoapps/lang_pref/views.py:
```diff
@@ -0,0 +1,22 @@
+from django.http import HttpResponse
+
+
+@csrf_exempt
```
what is the reason for `@csrf_exempt` here I believe we are sending CSRFToken for this.

asadiqbal08 replied:
@ziafazal I think, we can exempt this view from csrf, It is just updating the language session for unauthenticated users. I know that, we are sending csrf token in headers but it has been changed after user sign in. The csrf token is rotated when the user logs in. This was added as a security measure in Django 1.5.2 [release notes](https://docs.djangoproject.com/en/1.8/releases/1.5.2/).
please share your thoughts.

asadiqbal08 replied:
otherwise we need to fetch it from cookies.

ziafazal commented on common/test/acceptance/tests/studio/test_studio_home.py:
```diff
@@ -139,3 +144,36 @@ def test_tab_requires_staff(self):
+            u'English'
+        )
+
+        select_option_by_text(language_selector, 'Dummy Language (Esperanto)')
```
nit: may be we can put `Dummy Language (Esperanto)` in a variable.

ziafazal commented:
@asadiqbal08 completed my first round of review.

mattdrayer commented:
@andy-armstrong who would be the best person to take a quick pass through this PR and comment on the CSS?

asadiqbal08 commented:
@ziafazal I have updated the PR with yours suggestions.

asadiqbal08 commented on common/djangoapps/lang_pref/middleware.py:
```diff
@@ -21,20 +21,26 @@ def process_request(self, request):
+                if user_pref in system_released_languages:
+                    request.session[LANGUAGE_SESSION_KEY] = user_pref
+                else:
+                    delete_user_preference(request.user, LANGUAGE_KEY)
```
@mattdrayer fyi

ziafazal commented:
@asadiqbal08 LGTM :+1:

mattdrayer commented:
:+1: for me as well -- @AlasdairSwan if you have some time can you take a quick look at the CSS changes in this PR?

AlasdairSwan commented on common/static/js/src/lang_edx.js:
```diff
@@ -18,32 +15,49 @@ var edx = edx || {},
+                        if (is_user_authenticated) {
+                            // User language preference has been set successfully
+                            // Now submit the form in success callback.
+                            $("#language-settings-form").submit();
```
Nitpick: Please be consistent with quotes. The rest of the file uses single quotes so please update these double quotes to be single too.

AlasdairSwan commented on common/static/js/src/lang_edx.js:
```diff
@@ -18,32 +15,49 @@ var edx = edx || {},
+            submitAjaxRequest: function(language, url, callback) {
+
+                $.ajax({
+                    type: "PATCH",
```
Nitpick: Please be consistent with quotes. The rest of the file uses single quotes so please update these double quotes to be single too.

AlasdairSwan commented on common/static/js/src/lang_edx.js:
```diff
@@ -18,32 +15,49 @@ var edx = edx || {},
+                    data: JSON.stringify({'pref-lang': language}) ,
+                    url: url,
+                    dataType: 'json',
+                    contentType: "application/merge-patch+json",
```
Nitpick: Please be consistent with quotes. The rest of the file uses single quotes so please update these double quotes to be single too.

AlasdairSwan commented on lms/static/sass/shared/_header.scss:
```diff
@@ -119,7 +119,10 @@ header.global {
     .settings-language-form {
-      margin-top: 4px;
+      margin: 4px;
+      select {
```
Please use a class here instead of the element selector.

AlasdairSwan commented on cms/static/sass/elements/_header.scss:
```diff
@@ -43,8 +43,19 @@
+    margin: 0 10px 0 5px;
+    vertical-align: sub;
+
+    select {
```
Please use a class here instead of the element selector.

AlasdairSwan commented on cms/static/sass/elements/_header.scss:
```diff
@@ -208,6 +219,13 @@
+
+    .settings-language-form {
+      margin-top: 4px;
+      select {
```
Please use a class here instead of the element selector.

asadiqbal08 commented:
@mattdrayer can we go with merge ?

asadiqbal08 commented:
jenkins run lettuce

mattdrayer commented:
:+1:
